File: contrib/garmin-sidecar/wasm/garmin_ble_wasm.c
#include "../c/garmin_ble_core.h"

#include <string.h>

#ifdef __EMSCRIPTEN__
#include <emscripten/emscripten.h>
#define WASM_EXPORT EMSCRIPTEN_KEEPALIVE
#else
#define WASM_EXPORT
#endif

static garmin_ble_core_t g_core;
static uint8_t g_download_buffer[GARMIN_DOWNLOAD_MAX];
static garmin_packet_t g_last_write;
static garmin_event_t g_last_event;
static uint8_t g_last_payload[GARMIN_PACKET_MAX];

WASM_EXPORT void garmin_wasm_init(int prefer_reliable, unsigned int max_packet_size)
{
	garmin_ble_core_init(&g_core, prefer_reliable, max_packet_size);
	garmin_ble_core_set_download_buffer(&g_core, g_download_buffer, sizeof(g_download_buffer));
	memset(&g_last_write, 0, sizeof(g_last_write));
	memset(&g_last_event, 0, sizeof(g_last_event));
}

WASM_EXPORT int garmin_wasm_start(void)
{
	return garmin_ble_core_start(&g_core);
}

WASM_EXPORT int garmin_wasm_on_notification(const uint8_t *data, unsigned int size)
{
	garmin_event_t event;
	garmin_status_t rc = garmin_ble_core_on_notification(&g_core, data, size, &event);
	if (rc != GARMIN_OK) {
		memset(&g_last_event, 0, sizeof(g_last_event));
		g_last_event.type = GARMIN_EVENT_ERROR;
		g_last_event.status = rc;
		return rc;
	}

	g_last_event = event;
	if (event.payload != 0 && event.payload_size <= sizeof(g_last_payload)) {
		memcpy(g_last_payload, event.payload, event.payload_size);
		g_last_event.payload = g_last_payload;
	}
	return GARMIN_OK;
}

WASM_EXPORT int garmin_wasm_send_gfdi(unsigned int message_id, const uint8_t *payload, unsigned int payload_size)
{
	return garmin_ble_core_send_gfdi(&g_core, (uint16_t) message_id, payload, payload_size);
}

WASM_EXPORT int garmin_wasm_request_filter(void)
{
	return garmin_ble_core_request_filter(&g_core);
}

WASM_EXPORT int garmin_wasm_request_directory(void)
{
	return garmin_ble_core_request_directory(&g_core);
}

WASM_EXPORT int garmin_wasm_request_file(unsigned int file_index, unsigned int data_type, unsigned int sub_type, unsigned int file_size)
{
	garmin_directory_entry_t entry;
	memset(&entry, 0, sizeof(entry));
	entry.file_index = (uint16_t) file_index;
	entry.data_type = (uint8_t) data_type;
	entry.sub_type = (uint8_t) sub_type;
	entry.file_size = file_size;
	return garmin_ble_core_request_file(&g_core, &entry);
}

WASM_EXPORT int garmin_wasm_take_write(void)
{
	return garmin_ble_core_take_write(&g_core, &g_last_write);
}

WASM_EXPORT const uint8_t *garmin_wasm_write_ptr(void)
{
	return g_last_write.data;
}

WASM_EXPORT unsigned int garmin_wasm_write_size(void)
{
	return (unsigned int) g_last_write.size;
}

WASM_EXPORT int garmin_wasm_event_type(void)
{
	return g_last_event.type;
}

WASM_EXPORT unsigned int garmin_wasm_event_service(void)
{
	return g_last_event.service;
}

WASM_EXPORT unsigned int garmin_wasm_event_message_id(void)
{
	return g_last_event.message_id;
}

WASM_EXPORT const uint8_t *garmin_wasm_event_payload_ptr(void)
{
	return g_last_event.payload;
}

WASM_EXPORT unsigned int garmin_wasm_event_payload_size(void)
{
	return (unsigned int) g_last_event.payload_size;
}

WASM_EXPORT unsigned int garmin_wasm_event_offset(void)
{
	return g_last_event.offset;
}

WASM_EXPORT unsigned int garmin_wasm_event_total_size(void)
{
	return g_last_event.total_size;
}

WASM_EXPORT unsigned int garmin_wasm_event_file_index(void)
{
	return g_last_event.directory_entry.file_index;
}

WASM_EXPORT unsigned int garmin_wasm_event_file_data_type(void)
{
	return g_last_event.directory_entry.data_type;
}

WASM_EXPORT unsigned int garmin_wasm_event_file_sub_type(void)
{
	return g_last_event.directory_entry.sub_type;
}

WASM_EXPORT unsigned int garmin_wasm_event_file_size(void)
{
	return g_last_event.directory_entry.file_size;
}

File: contrib/garmin-sidecar/wasm/garmin_ble_wasm.js
export function createGarminBleCore(Module) {
  const u8 = () => Module.HEAPU8;

  function readBytes(ptr, size) {
    return new Uint8Array(u8().slice(ptr, ptr + size));
  }

  function withBytes(bytes, fn) {
    const input = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
    const ptr = Module._malloc(input.length || 1);
    try {
      u8().set(input, ptr);
      return fn(ptr, input.length);
    } finally {
      Module._free(ptr);
    }
  }

  return {
    init({ reliable = true, maxPacketSize = 20 } = {}) {
      Module._garmin_wasm_init(reliable ? 1 : 0, maxPacketSize);
    },

    start() {
      const rc = Module._garmin_wasm_start();
      if (rc !== 0) throw new Error(`Garmin start failed: ${rc}`);
      return this.takeWrites();
    },

    onNotification(bytes) {
      const rc = withBytes(bytes, (ptr, size) => Module._garmin_wasm_on_notification(ptr, size));
      if (rc !== 0) throw new Error(`Garmin notify failed: ${rc}`);
      return {
        event: this.lastEvent(),
        writes: this.takeWrites(),
      };
    },

    sendGfdi(messageId, payload = new Uint8Array()) {
      const rc = withBytes(payload, (ptr, size) => Module._garmin_wasm_send_gfdi(messageId, ptr, size));
      if (rc !== 0) throw new Error(`Garmin GFDI send failed: ${rc}`);
      return this.takeWrites();
    },

    requestFilter() {
      const rc = Module._garmin_wasm_request_filter();
      if (rc !== 0) throw new Error(`Garmin filter request failed: ${rc}`);
      return this.takeWrites();
    },

    requestDirectory() {
      const rc = Module._garmin_wasm_request_directory();
      if (rc !== 0) throw new Error(`Garmin directory request failed: ${rc}`);
      return this.takeWrites();
    },

    requestFile({ fileIndex, dataType = 128, subType = 4, fileSize = 0 }) {
      const rc = Module._garmin_wasm_request_file(fileIndex, dataType, subType, fileSize);
      if (rc !== 0) throw new Error(`Garmin file request failed: ${rc}`);
      return this.takeWrites();
    },

    takeWrites() {
      const writes = [];
      while (Module._garmin_wasm_take_write()) {
        writes.push(readBytes(Module._garmin_wasm_write_ptr(), Module._garmin_wasm_write_size()));
      }
      return writes;
    },

    lastEvent() {
      const payloadSize = Module._garmin_wasm_event_payload_size();
      const payloadPtr = Module._garmin_wasm_event_payload_ptr();
      return {
        type: Module._garmin_wasm_event_type(),
        service: Module._garmin_wasm_event_service(),
        messageId: Module._garmin_wasm_event_message_id(),
        offset: Module._garmin_wasm_event_offset(),
        totalSize: Module._garmin_wasm_event_total_size(),
        file: {
          fileIndex: Module._garmin_wasm_event_file_index(),
          dataType: Module._garmin_wasm_event_file_data_type(),
          subType: Module._garmin_wasm_event_file_sub_type(),
          fileSize: Module._garmin_wasm_event_file_size(),
        },
        payload: payloadPtr && payloadSize ? readBytes(payloadPtr, payloadSize) : new Uint8Array(),
      };
    },
  };
}
